aptoide-installer/src/main/java/cm/aptoide/pt/installer/fetch/FetchDownloader.kt

```diff
@@ -0,0 +1,215 @@
+        .setHttpDownloader(
+          object : OkHttpDownloader(
+            okHttpClient,
+            fileDownloaderType = Downloader.FileDownloaderType.PARALLEL
+          ) {
+            override fun verifyContentHash(
+              request: Downloader.ServerRequest,
+              hash: String
+            ): Boolean {
+              return request.fileUri.toFile().checkMd5(contentHashes[request.id] ?: hash)
+            }
+
+            override fun getHeadRequestMethodSupported(request: Downloader.ServerRequest): Boolean {
+              return true
+            }
+          })
+        .build()
+    )
+  }
+
+  @Suppress("OPT_IN_USAGE")
+  override fun download(
+    packageName: String,
+    installPackageInfo: InstallPackageInfo,
+  ): Flow<Int> = installPackageInfo.run {
+    flow<Unit> {
+      if (installPackageInfo.hasObb()) {
+        installPermissions.checkIfCanWriteExternal()
+      }
+      installPermissions.checkIfCanInstall()
+      emit(Unit)
+    }.flatMapLatest {
+      downloadFiles(packageName, versionCode, installationFiles)
+    }.map {
+      coroutineContext.ensureActive()
+      (it * 100).toInt()
+    }
+  }
+    .also { InstallerWorker.enqueue(context) }
+
+  private fun downloadFiles(
+    packageName: String,
+    versionCode: Long,
+    installationFiles: Set<InstallationFile>
+  ): Flow<Double> {
+    val totalSize = installationFiles.sumOf { it.fileSize }.toDouble()
+    val fetch = Fetch.Impl.getDefaultInstance()
+
+    val destinationDir = File(downloadsPath, packageName).apply {
+      if (!exists()) {
+        mkdirs().let {
+          if (!it) throw IllegalStateException("Can't create download folder: $downloadsPath/$packageName")
+        }
+      }
+    }
+
+    val requests = installationFiles.mapNotNull { file ->
+      File(destinationDir, file.name)
+        .takeUnless { it.checkMd5(file.md5) }
+        ?.let {
+          Request(file.url, it.toUri())
+            .apply {
+              priority = Priority.NORMAL
+              networkType = NetworkType.ALL
+              groupId = packageName.asGroupId()
+              addHeader(VERSION_CODE, versionCode.toString())
+              addHeader(PACKAGE, packageName)
+              addHeader(FILE_TYPE, file.type.toString())
+            }
+            .also {
+              contentHashes.put(it.id, file.md5)
+            }
+        }
+    }
+
+    val downloads = requests.map { it.id }.toMutableSet()
+    val downloadsProgresses = requests.associate { it.id to 0L }.toMutableMap()
+
+    return callbackFlow<Double> {
+      try {
+        trySend(0.0)
+
+        if (downloads.isEmpty()) {
+          trySend(1.0)
+          close()
+        }
+
+        fetch.enqueue(requests) { enqueuedRequests ->
+          enqueuedRequests.find { it.second != Error.NONE }?.let {
+            close(it.second.throwable ?: IllegalStateException("Error enqueuing requests"))
+          }
+        }
+
+        val listener = object : FetchFileDownloadListener {
```

Comment:
Why we didnt use the cancel or remove or delete callback ? Wouldnt they be necessary? I know we are handling it ourselves but i am just asking to be sure if there are more possible ways for it to be called

Reply:
I don't think there is any action that needs to be done after a file is removed/deleted so those callbacks are not necessary. Regarding the cancel callback, since the cancellation is only caught on the flow's completion, both the callbackFlow and listener won't be active anymore